### PR TITLE
Add data channel transferable algorithm

### PR DESCRIPTION
**Pull request by youennf**
Fixes https://github.com/w3c/webrtc-extensions/issues/64

***
<a href="https://pr-preview.s3.amazonaws.com/youennf/webrtc-extensions/pull/78.html" title="Last updated on Jun 4, 2021, 9:29 AM UTC (40c0478)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/webrtc-extensions/78/e706a8f...youennf:40c0478.html" title="Last updated on Jun 4, 2021, 9:29 AM UTC (40c0478)">Diff</a>

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -920,5 +920,69 @@ <h3>
...
+      This allows sending and receiving messages outside the context the connection was created, for instance in workers or third-party iframes.</p>
+    <div>
+      <p>The WebIDL changes are the following:
+      <pre class="idl"
+>[Exposed=(Window,Worker), Transferable]
+partial interface RTCDataChannel {
+};</pre>
+    </div>
+    <div>
+      <p>At creation of a {{RTCDataChannel}} object, called <var>channel</var>, run the following steps:</p>
```
Comment (jan-ivar) [inline index.html:937]: We should reference https://w3c.github.io/webrtc-pc/#dfn-create-an-rtcdatachannel explicitly here.

**Comment by youennf** [comment]
https://w3c.github.io/webrtc-identity/#creating-an-offer could be used to do better monkey patching

File: index.html (hunk 1 above)
Comment (youennf) [reply index.html]: Done

File: index.html [hunk 2]
```diff
@@ -920,5 +920,69 @@ <h3>
...
+      <ol>
+        <li><p>Set <var>channel</var>.`[[IsTransferable]]` to <code>false</code>.</p></li>
+      </ol>
+    </div>
+    <div>
+      <p>The {{RTCDataChannel}} <a data-cite="!HTML/#transfer-steps">transfer steps</a>, given <var>value</var> and <var>dataHolder</var>, are:</p>
+      <ol>
+        <li><p>If <var>value</var>.`[[IsTransferable]]` is <code>false</code>, throw a "DataCloneError" DOMException.</p></li>
+        <li><p>Set <var>dataHolder</var>.`[[ReadyState]]` to <var>value</var>.<a href="WEBRTC#dfn-readystate">`[[ReadyState]]`</a>.</p></li>
+        <li><p>Set <var>dataHolder</var>.`[[BufferedAmount]]` to <var>value</var>.<a href="WEBRTC#dfn-bufferedamount">`[[BufferedAmount]]`</a>.</p></li>
```
Comment (alvestrand) [inline index.html:957]: I dont' think we should transfer bufferedAmount, because that value could change.

File: index.html [hunk 3]
```diff
@@ -920,5 +920,69 @@ <h3>
...
+        <li><p>Set <var>dataHolder</var>.`[[DataChannelId]]` to <var>value</var>.<a href="WEBRTC#dfn-datachannelid">`[[DataChannelId]]`</a>.</p></li>
+        <li><p>Set <var>dataHolder</var>.`[[underlyingDataTransport]]` to <var>value</var> <a href="WEBRTC#dfn-data-transport">underlying data transport</a>.</p></li>
+        <li><p>Set <var>value</var>.`[[IsTransferable]]` to <code>false</code>.</p></li>
+        <li><p>Set <var>value</var>.`[[ReadyState]]` to <a data-cite="WEBRTC/#dom-rtcdatachannelstate-closed">"closed"</a>.</p></li>
+      </ol>
+    </div>
+    <div><p>The {{RTCDataChannel}} <a data-cite="!HTML/#transfer-receiving-steps">transfer-receiving steps</a>, given <var>dataHolder</var> and <var>channel</var>, are:</p>
+      <ol>
+        <li><p>Initialize <var>channel</var>.<a href="WEBRTC#dfn-readystate">`[[ReadyState]]`</a> to <var>dataHolder</var>.`[[ReadyState]]`.</p></li>
+        <li><p>Initialize <var>channel</var>.<a href="WEBRTC#dfn-bufferedamount">`[[BufferedAmount]]`</a> to <var>dataHolder</var>.`[[BufferedAmount]]`.</p></li>
```
Comment (alvestrand) [inline index.html:973]: I think this should instead say "Initialize BufferedAmount by querying the underlying data transport's state".

File: index.html [hunk 4]
```diff
@@ -920,5 +920,69 @@ <h3>
               </dd>
             </dl>
 </section>
+<section>
+    <h3>
+      {{RTCDataChannel}} extensions
+    </h3>
+    <p>This document extends {{RTCDataChannel}} by making it <a data-cite="!HTML/#transferable-objects">transferable</a>.</p>
+      This allows sending and receiving messages outside the context the connection was created, for instance in workers or third-party iframes.</p>
```
Comment (alvestrand) [inline index.html:928]: I think this paragraph can be extended a little bit to give more context. In particular, something like:

An RTCDataChannel can be transferred immediately after being created. As soon as incoming data may be present on the RTCDataChannel, it is no longer transferable.

It is possible to call send() on a DataChannel before transferring it.

File: index.html [hunk 5]
```diff
@@ -920,5 +920,69 @@ <h3>
...
+};</pre>
+    </div>
+    <div>
+      <p>The <a data-cite="WEBRTC#dfn-create-an-rtcdatachannel">create an RTCDataChannel</a> algorithm is updated by adding the following steps after step 4 of the original algorithm:</p>
+      <ol>
+        <li><p>Initialize <var>channel</var>.`[[IsTransferable]]` to <code>true</code>.</p></li>
+        <li><p>Queue a task to run the following step:</p>
+          <ol>
+            <li><p>Set <var>channel</var>.`[[IsTransferable]]` to <code>false</code>.</p></li>
+          </ol>
```
Comment (alvestrand) [inline index.html:943]: Suggest adding a paragraph saying "This task must be guaranteed to run before any incoming data is presented to "channel"." It is not obvious that the implementation has to enforce that property; in theory there could be ondata events queued behind the onopen event before the onopen event gets started; this transfer algorithm has to ban that implementation strategy.

**Review by alvestrand — COMMENTED** [review]
Looks good, have some comments I think we should address to make requirements clearer.

File: index.html (hunk 2 above)
Comment (youennf) [reply index.html:957]: That value should always be 0, otherwise channel is not transferable

File: index.html (hunk 4 above)
Comment (youennf) [reply index.html:928]: For simplicity, calling send() will not make the data channel transferable anymore.

File: index.html (hunk 2 above)
Comment (alvestrand) [reply index.html:957]: Then there's no need to transfer it.

Good that you added the prohibition against send() explicitly - I was reviewing the version before that change.

File: index.html (hunk 5 above)
Comment (youennf) [reply index.html:943]: Sounds fair, I'll add something

File: index.html (hunk 3 above)
Comment (alvestrand) [reply index.html:975]: Or even simpler: "set it to zero".

File: index.html (hunk 2 above)
Comment (alvestrand) [reply index.html:959]: Or even simpler: Don't include it in the dataHolder.

File: index.html (hunk 4 above)
Comment (aboba) [reply index.html:928]: Agree that allowing calling `send()` should probably be disallowed for simplicity. I'm more worried about data that may be in the receive pipeline.  Data can be sent by the other peer as soon as the SctpTransport is established and the data channel is open. Why force the queued data to be thrown away?

Comment (youennf) [reply index.html:940]: As discussed during the call, we make sure no message data is lost in the transfer.
This is explicitly stated in the latest PR.

File: index.html (hunk 3 above)
Comment (youennf) [reply index.html]: Removed the line, according yesterday's discussion

File: index.html (hunk 2 above)
Comment (youennf) [reply index.html]: Done

**Pull request merged**